[#1119] improvement: Explicitly throw BUFFER_LIMIT_OF_HUGE_PARTITION instead of NO_BUFFER

lifeSo opened the pull request:
### What changes were proposed in this pull request?

Explicitly throw BUFFER_LIMIT_OF_HUGE_PARTITION instead of NO_BUFFER

### Why are the changes needed?


Fix: # https://github.com/apache/incubator-uniffle/issues/1119

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

test on prd and unit test case

zuston commented on server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java:
```diff
@@ -458,31 +459,37 @@ public long requireBuffer(
     ShuffleTaskInfo shuffleTaskInfo = shuffleTaskInfos.get(appId);
     if (null == shuffleTaskInfo) {
-      return RequireBufferStatusCode.NO_REGISTER.statusCode();
+      LOG.error("Find not registered app, appId: {}, shuffleId: {}", appId, shuffleId);
```
How about `no such app is registered. appId: {}, shuffleId: {}`

zuston commented on server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java:
```diff
@@ -458,31 +459,37 @@ public long requireBuffer(
           appId, shuffleId, partitionId, partitionUsedDataSize)) {
         ShuffleServerMetrics.counterTotalRequireBufferFailedForHugePartition.inc();
-        return RequireBufferStatusCode.NO_BUFFER.statusCode();
+        String errorMessage =
```
Why not throwing the `NoBufferForHugePartitionException` ?

zuston commented on server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java:
```diff
@@ -458,31 +459,37 @@ public long requireBuffer(
-        return RequireBufferStatusCode.NO_BUFFER.statusCode();
+        String errorMessage =
+            String.format(
+                "No Buffer For Huge Partition, appId: %s, shuffleId: %s, partitionIds: %s, partitionUsedDataSize: %s",
```
How about `Huge partition is limited to writing. appId: %s, shuffleId: %s, partitionIds: %s, partitionUsedDataSize: %s` ?

zuston commented on server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java:
```diff
@@ -393,28 +394,30 @@ public void requireBuffer(
+      requireBufferId = -1;
       status = StatusCode.NO_REGISTER;
       ShuffleServerMetrics.counterTotalRequireBufferFailed.inc();
+    } catch (NoBufferException e) {
```
How to handle `NoBufferForHugePartitionException,` I don't see these code/.

zuston reviewed (commented):
Could you add some tests for huge partition no buffer exception ?

lifeSo commented:
> Could you add some tests for huge partition no buffer exception ?

Ok, I added sendDataAndRequireBufferTest() test case in ShuffleServerGrpcTest. Mainly:
![image](https://github.com/apache/incubator-uniffle/assets/9440617/a38deb1e-284b-404c-be60-0ef56b734f49)

zuston commented on server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java:
```diff
@@ -393,28 +395,30 @@ public void requireBuffer(
+      requireBufferId = -1;
       status = StatusCode.NO_REGISTER;
       ShuffleServerMetrics.counterTotalRequireBufferFailed.inc();
+    } catch (NoBufferException | NoBufferForHugePartitionException e) {
```
This looks weird, why not introducing new stateCode of `NoBufferForHugePartitionException`

zuston reviewed (commented):
I hope the memory message of being limited by huge partition should be shown in client side rather than only having one exception in server

jerqi commented on common/src/main/java/org/apache/uniffle/common/rpc/StatusCode.java:
```diff
@@ -27,13 +27,14 @@ public enum StatusCode {
-  TIMEOUT(7),
-  ACCESS_DENIED(8),
-  INVALID_REQUEST(9),
+  NO_BUFFER_FOR_HUGE_PARTITION(3),
```
You should consider the compatility. 
```
NO_BUFFER_FOR_HUGE_PARTITION(10)
```

zuston replied:
+1. should be 11

zuston commented on integration-test/common/src/test/java/org/apache/uniffle/test/IntegrationTestBase.java:
```diff
@@ -58,6 +58,8 @@ public abstract class IntegrationTestBase extends HadoopTestBase {
   protected static final int JETTY_PORT_1 = 19998;
   protected static final int JETTY_PORT_2 = 20040;
   protected static final String COORDINATOR_QUORUM = LOCALHOST + ":" + COORDINATOR_PORT_1;
+  protected static final String SHUFFLE_SERVER_METRICS_URL =
```
why need this?

zuston commented on server/src/main/java/org/apache/uniffle/server/ShuffleServerMetrics.java:
```diff
@@ -80,7 +80,9 @@ public class ShuffleServerMetrics {
   public static final String LOCAL_DISK_PATH_LABEL_ALL = "ALL";
   private static final String TOTAL_REQUIRE_BUFFER_FAILED = "total_require_buffer_failed";
-  private static final String TOTAL_REQUIRE_BUFFER_FAILED_FOR_HUGE_PARTITION =
+  private static final String TOTAL_REQUIRE_BUFFER_FAILED_FOR_NO_REGISTER =
```
this is not related with this PR. Right?

zuston commented on internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java:
```diff
@@ -223,7 +223,8 @@ public long requirePreAllocation(
     final int backOffBase = 2000;
-    while (rpcResponse.getStatus() == RssProtos.StatusCode.NO_BUFFER) {
+    while (rpcResponse.getStatus() == RssProtos.StatusCode.NO_BUFFER
+        || rpcResponse.getStatus() == RssProtos.StatusCode.NO_BUFFER_FOR_HUGE_PARTITION) {
```
Could you expose the concrete status code into log message?

zuston replied:
ping @lifeSo

jerqi commented:
Could you reply all the comments from zuston?

lifeSo commented:
> **[zuston](/zuston) ** reviewed [ Dec 28, 2023 ](#pullrequestreview-1797951840)
> [ View reviewed changes ](/apache/incubator-uniffle/pull/1393/files/f89872ca15ab704b0cbe3b5c23382f8ce6d1d16f)



> Could you reply all the comments from zuston?

@jerqi  I checked and find one place miss modify, and I corrected it. 
I thin all comments is reply and corrected.

jerqi commented:
> > **[zuston](/zuston) ** reviewed [ Dec 28, 2023 ](#pullrequestreview-1797951840)
> > [ View reviewed changes ](/apache/incubator-uniffle/pull/1393/files/f89872ca15ab704b0cbe3b5c23382f8ce6d1d16f)
> 
> > Could you reply all the comments from zuston?
> 
> @jerqi I checked and find one place miss modify, and I corrected it. I thin all comments is reply and corrected.

You should reply the comment instead of marking the comment resolved. The reply is the your solution. Maybe it will be very simple like `fixed`. Let reviewer mark the comment resolved.

zuston commented:
#1425 Merged. And close this.
